stats/view/worker.go

```diff
@@ -233,6 +233,7 @@ func (w *worker) reportView(v *viewInternal, now time.Time) {
 		e.ExportView(viewData)
 	}
 	exportersMu.Unlock()
+	v.clearRows()
```

Comment:
This will affect sum/count and bucket count in Distribution. It will also affect count aggregation which is expected to be cumulative.

I understand the problem of min and max that you are trying to solve. Reseting just min and max might be the solution.

Reply:
@rghetia Thanks for the review :) 

So I don't think it's just the min/max. I'm not sure about the sum/count in Distribution just yet, but for the Count-Aggregation, it should probably also be reset and let the Exporter be responsible for the cumulativeness side of things. 

The reason I say that is because if the server restarts for whatever reason, then the count is completely gone. Furthermore, if the Count is displayed on a timeseries graph, then a lot of these values will be falsely duplicated. 

For example, imagine a server receives 10 requests per second..after 10 seconds, the graph will say there are 100 requests in the last *second* (which is wrong, there were only 10) and if the serve restarts you will see a sharp drop from 100 back to 10. 

If we collect the "count per interval", then that should solve the issue. 

Would also love to know more about which data should never be reset, *even* if the process restarts and the data gets "forced" to be reset.

Reply:
Distributions and Counters have Start-time. Start-time lets backend detect the restart. 
It also lets it calculate the rate correctly. 
Example:
```
At time T0: Application starts 
T1: First request served. Start-time is set to T1
T2: 100 request served. 
T3: 200 request served. Export enabled.
T4: 300 request served. First export happens, Count = 300, Start-time = T1. Rate=300/(T4-T1)
T5: 400 request served.
T6: Application Restarts
T7: First request served after restart, Start-time= T7
T8: 100 request served. Export count = 100, Start-time = T7, Rate = 100/(T8-T7)
```
Resetting all metrics would be a behavior change and would have an adverse consequence on some backends.

Reply:
@rghetia awesome. However, from what I can see there's no way for a backend to determine that: 

The CountData only has access to `Value` and on idea of start time: https://github.com/census-instrumentation/opencensus-go/blob/master/stats/view/aggregation_data.go#L42

In your example, the Exporter does not have access to `T1` (does it?), it can have access to `T0` by just putting `func init() { t1 = time.Now() }` I imagine but that's not accurate. 

Furthermore, dividing Count by (Tnow-T1) will just make an even distribution across time, and not the _real_ distribution. For example, if the first minute  we got 300 requests, but the next minute we got 10 requests, it will not show that accurately no? 

Finally, do you have an example of an exporter doing this time calculation? 

Thanks again :)

Reply:
startTime per view is stored [here](https://github.com/census-instrumentation/opencensus-go/blob/df6e2001952312404b06f5f6f03fcb4aec1648e5/stats/view/worker.go#L44)

There are two interface for exporters to get the data
1. [ExportView(vd *ViewData)](https://github.com/census-instrumentation/opencensus-go/blob/3b5a343282fe4b4fccdb0f24cbd1d7169d20858a/stats/view/export.go#L35)
starttime for ViewData is set [here](https://github.com/census-instrumentation/opencensus-go/blob/df6e2001952312404b06f5f6f03fcb4aec1648e5/stats/view/worker.go#L227)
2. [ExportMetric(context.Context, []*metricdata.Metric)](https://github.com/census-instrumentation/opencensus-go/blob/3b5a343282fe4b4fccdb0f24cbd1d7169d20858a/metric/metricexport/export.go#L25)
startTime for Metric is retrieved [here](https://github.com/census-instrumentation/opencensus-go/blob/3b5a343282fe4b4fccdb0f24cbd1d7169d20858a/stats/view/worker.go#L261)
and set [here](https://github.com/census-instrumentation/opencensus-go/blob/df6e2001952312404b06f5f6f03fcb4aec1648e5/stats/view/view_to_metric.go#L115)

Regarding even distribution - If backend has individual data points then it can correctly represent the rate for each time window. For example
T0: Count = 300
T1: Count = 400
T2: Count = 410
Rate for the window T1-T0 = (400-300)/(T1-T0)
Rate for the window T2-T1 = (410-400)/(T2-T1)

ViewData and Metric both have end time as well. So the backend can compute the rate accordingly. 

I'll post a screenshot shortly.

Reply:
![image](https://user-images.githubusercontent.com/8052115/68438735-f5849980-0179-11ea-8052-6711cd2d73e5.png)

Reply:
@rghetia 

>If backend has individual data points then it can correctly represent the rate for each time window

What do you mean by individual data points? Does the Exporter have to keep track of "previous ExportData" metrics? 

If that's true, then an Exporter wouldn't even need to care about the time, it can just take the "last count value" and subtract that from the current value and that's the right number for every ExportView invocation, right? 

For example, 

T0: Process starts
T1: Count = 100 -- Real Count is  100 because (100-0 = 100)
T2: Count = 150 -- Real Count is 150-100 because (150-100 = 50)

Therefore, at T1 count is 100, and at T2 count is 50. 

Please let me know if my understanding is correct. 

Lastly, can you tell me which of the data here needs to be "reset" and which must stay the same? https://github.com/census-instrumentation/opencensus-go/blob/master/stats/view/aggregation_data.go#L118:L129

Thanks again 🎉

Reply:
> What do you mean by individual data points? Does the Exporter have to keep track of "previous ExportData" metrics?
> 
> If that's true, then an Exporter wouldn't even need to care about the time, it can just take the "last count value" and subtract that from the current value and that's the right number for every ExportView invocation, right?

Exporter doesn't keep any state. It simply sends the data to the backend. 
The start-time and end-time are both important from the backend perspective.
Start-time is used to detect reset. It is also used to estimate if initial data points were missed.
End-time is important to accurately compute the rate between two data points.
Example:

| Time | Value | Start Time | Exported | Rate | Comments |
| ----- | ----- | ---------- | --------- | --- | ---------- |
| T0+60s | 600 | T0 | No |  |  |
| T0+120s | 1200 | T0 | No |  |  |
| T0+180s | 1800 | T0 | Yes | 10/s  | Rate is still computed even though the first two data points are missing. Without start-time it would not be able to compute the rate |
| T0+240s | 2100 | T0 | Yes | 5/s  | Actual rate from time T0+180 to T0+240s  |
| T0+560s | 600 | T0+500 | Yes | 10/s  | Reset Detected T0 != T0+500 |
| T0+620s | 1200 | T0+500 | No | 10/s  | Lost for whatever reason |
| T0+680s |1800 | T0+500 | Yes | 10/s  | Rate is computed properly even though previous point was missing. Without the end-time it would be able to calculate the rate. Alternatively backend could record time on receiving data point but that would not be as accurate as the time at the observation point |

> 
> Lastly, can you tell me which of the data here needs to be "reset" and which must stay the same? https://github.com/census-instrumentation/opencensus-go/blob/master/stats/view/aggregation_data.go#L118:L129
> 
> I imagine the min/max definitely need to be reset, probably the Mean as well? But I'm not sure about the rest of the struct fields here.

you can only reset min/max.

Reply:
> Reset Detected T0 != T0+500

That's interesting. However, from all the exporters I'm looking at, *none* of them even mention the StartTime let alone pass it back to their respective backends...Are all of these backends incorrect? More likely that I'm missing something here, but the ones I'm looking at are [Prometheus](https://github.com/census-ecosystem/opencensus-go-exporter-prometheus) and [Datadog](https://github.com/DataDog/opencensus-go-exporter-datadog). 

I'm also looking at the getting started guide [here](https://opencensus.io/quickstart/go/metrics/) and you can see that all the data goes up cumulatively, but if I reset the process, then the data drops and start over, so it's not automatically doing any time calculation the way you explained it above. Maybe I'm missing a configuration or a knob somewhere? 

On the other hand, I think we're talking about two different models: the OpenCensus pull model that you very kindly explained. But on the other hand, the model DataDog operates on is the [StatsD Model](https://github.com/b/statsd_spec) which from what I gather is quite trivial in that it only takes one number at a specific time  and that's the end of the story. Therefore, the only way to solve the problem is to actually make the Exporter keep **state**. At my company we have a contact with DataDog so I'll try to reach out to them about this. 

In any case, thank you @rghetia  for the thorough explanation 💯 💯

Reply:
@rghetia I'll send a fix for min/max but before that I'd love to confirm that the exporters I mentioned above are indeed not working as intended. Thanks!

Reply:
yes prometheus doesn't handle the reset because it doesn't have start-time.
[stackdriver](https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver) does use the start-time and handles the reset.
datadog seems to convert everything to Gauge (except histograms). In Gauge there is no start-time.  
By definition Count() and Sum() aggregation are defined as cumulative so unfortunately it can't be reset. What you are asking for is a Delta() aggregator which provides delta between two reads.

Reply:
> yes prometheus doesn't handle the reset because it doesn't have start-time.

This means that in Prometheus, the RATE is only accurate if the process never restarts. 

Furthermore, the COUNT is never accurate since I can't imagine prometheus doing the more complex Delta calculations on their end that we both described in the comments above (re-iterated here: checking if the previous value is smaller or bigger to determine if the process falsely restarted in order to aggregate the count accurately) 

> By definition Count() and Sum() aggregation are defined as cumulative

This is where there's a misunderstanding between specs and libraries. According to the DataDog [docs](https://docs.datadoghq.com/developers/metrics/metrics_type/?tab=count#metric-type-definition):

> ***The COUNT metric submission type represents the number of events that occur in a defined time interval, otherwise known as a flush interval. This number of events can accumulate or decrease over time – it is not monotonically increasing.***

That said, you did mention the following: 

> datadog seems to convert everything to Gauge (except histograms)

In fact, even histograms (DistributionData) are converted to a Gauge: https://github.com/DataDog/opencensus-go-exporter-datadog/blob/master/stats.go#L77:L105

I understand that the exporter is treating the OpenCensus Count() type more of a DataDog GAUGE type than a DataDog COUNT type. But it's still incorrect because it's always gonna be the number since the process started, and it gets even worse when you are aggregating over multiple tags where each one has a different start time.


### MEAN
if Count() and Sum() are cumulative, what about the `DistributionData.Mean` field? Right now the Mean() is being calculated over the entire process, and similar to Min/Max, there's absolutely no way to calculate/fix process-restarts...right? Hopefully I'm wrong.

### DELTA 

> What you are asking for is a Delta() aggregator which provides delta between two reads.

To be exact, what I'm asking for is a Delta() aggregator which provides delta between two reads *as long as the process does not restart*, and if the process restarts then it needs to apply a more complex calculation to get the right delta (detailed above). 

By the way, I'm not really asking for the above, it sounds **very** cumbersome. Nonetheless, I am just trying to make OpenCensus work with 2 out of the 3 most commonly used stats exporters which both seem to be completely broken given our conversation so far. 

From an emotional perspective, I'm frustrated that this disconnect has been sitting there this entire time without anyone noticing. Like, did anyone test this at all? OpenCensus has docs and tutorials on both Prometheus and DataDog which is a clear endorsement of these exporters yet both seem quite broken to me. 

@rghetia thanks again for keeping up with all of this. If you'd like to discuss this over a video call for more real time discussion, I'm happy to do so. Otherwise, this thread has been very helpful as well. 

From a practical side of things, I'd like to solve this issue and move on. I'm currently using a fork of OpenCensus that is pinned at the commit of this PR but that's just a temporary workaround. 

### Moving Forward: 

So, let's see where we can go from here: 

We currently have 3 problems: 

1. Min/Max: solved since you're okay with merging this type of fix. 
2. Count(): we have two possible solutions (feel free to suggest others): 
2A. Create a new Delta type in OpenCensus 
2B. Make the exporters keep state and get the delta. 
3. Mean(): Right now, I'm not sure how I can get the mean of a histogram say between 12pm to 1pm (uninfluenced by data before 12pm and also unadulterated by process restarts between 12pm to 1pm)

Reply:
google hangout or zoom works. I am on PST. Ping me on gitter with time and a link.

Reply:
@rghetia I've updated the PR to reset min, max, mean (and also sqauredSumDev) as all of these values share the same problem we've discussed. 

Let me know if this looks good. 

Thanks again!

Reply:
@rghetia now that I think about, resetting Mean doesn't work without resetting all the buckets. Right? 

I have another idea, why don't I roll back to my original PR, but instead of forcing the restart, we just have it as a dynamic configuration? Something like `view.RestartAfterFlush(true)` that will set a config in the worker.. 

This way, StackDriver and exporters relying on Monotonicity won't break, but DataDog and other StatsD based backends will work as expected. 

Let me know!

Reply:
Mean and squaredSumDev are used in stackdriver (at least they are  exported to the backend). 
view.RestartAfterFlush(true) (not sure about the name though) is better but it has following caveats
- It will not work with [Derived metrics](https://github.com/census-instrumentation/opencensus-go/blob/aad2c527c5defcf89b5afab7f37274304195a6b2/metric/registry.go#L200). Derived metrics are simply read when metrics are exported. 
- There can only be one Exporter if new interface (Reader or IntervalReader) is used.

keeping a state in the exporter would work for all cases.

Reply:
>keeping a state in the exporter would work for all cases. 

Unfortunately, `DistributionData.Mean` cannot be reset without resetting `DistributionData.Count`, because the calculation looks like this in the code: `a.Mean = a.Mean + (v-a.Mean)/float64(a.Count)` -- in other words the count has to be reset or OpenCensus has to keep state of the previousCount itself. 

The only option I see that doesn't break StackDriver and others is to make the config an opt-in. The DataDog exporter doesn't use Derived metrics and the exporter can include in the documentation that it can only be used by itself...most people don't use multiple metrics exporters at the same time so that should be okay

 Thanks

Reply:
From the library perspective adding view.RestartAfterFlush seems like a hack. Are you only interested in reseting Distribution metrics after every export? If so, I am more inclined to add another aggregation type 'DistributionDelta' which will reset after every read. This would have to be added in the spec though and implemented across all languages.

I still think handling this in an exporter might be better at this point, especially if it is for distribution metrics only.

Reply:
@rghetia personally I'm interested in both `view.DistributionData` as well as `view.CountData` -- but just as An FYI, `view.SumData` also falls under this category. 

The Delta for view.CountData can be maintained in the exporter, but the Delta for `view.DistributionData.Mean` (as well as squaredSumDev) cannot be maintained without resetting view.DistributionData.Count. 

>I still think handling this in an exporter might be better at this point, especially if it is for distribution metrics only.

Can you confirm that that's impossible without resetting the DistributionData.Count? 

If so, we only have two options: 

1. Create Delta types for both Count and DistributionData and implement them everywhere
2. Add `view.DisableMonotonicity(true)` option to the configuration with the 2 caveats documented for it. This might be fine especially if OpenTelemetry will resolve the Delta issue natively?

Reply:
> The Delta for view.CountData can be maintained in the exporter, but the Delta for `view.DistributionData.Mean` (as well as squaredSumDev) cannot be maintained without resetting view.DistributionData.Count.

The way to do it is maintain the prev state for entire distribution. So calculation would be
```
exportDist.Count = currDist.Count - prevDistCount
exportDist.Mean = (currDist.Count * currDist.Mean - prevDist.Count * prevDist.Mean)/exportDist.Count
exportDist.Bucket[0].Count = currDist.Bucket[0].Count - prevDist.Bucket[0].Count
```
we do similar thing (maintaining state) in prometheus receiver ([challenge #5](https://github.com/open-telemetry/opentelemetry-collector/tree/master/receiver/prometheusreceiver#challenges-and-solutions))

Reply:
@rghetia ah, that sounds good. 

This is more of a question: how does the exporter keep track of previous data? Can it do it by each tag key-value individually? Or by the combination of each `*view.Row.Tags`?? 

I notice that the number of rows that are given to ExportView can vary for different reasons. 

Thank you!

Reply:
by combination of each view.Row.Tags. In prometheus receiver it does it by metric-name+sorted label-values(tag-values).

Reply:
@rghetia 

1. I'd love to see an example of that, I'll send you what I have soon as well. Thank you :pray: 
2. Is the ExportView expected to be called synchronously or concurrently? Just wondering if I should be guarding against checking-updating previous count/distribution data. 

Thanks again

Reply:
Also, I just reset the PR back to only resetting min/max :)

Reply:
Lastly, I'm not sure how we can calculate the difference in `SumOfSquaredDev` 🤔

Reply:
Example:

| metric | Labels | Count | Sum | Start-time| Delta Count | Delta Sum | comment |
|-------|--------|-------|------|-------------|------------|----------|-----------|
| M1 | K1=V11, K2=V21 | 10 | 100 | T0 | 0 | 0 | first report ignore |
| M1 | K1=V11, K2=V22 | 20 | 150 | T0 | 0 | 0 | first report ignore |
| M1 | K1=V11, K2=V11 | 20 | 200 | T0 | 10 | 100 | second report take delta |
| M1 | K1=V11, K2=V22 | 35 | 300 | T0 | 15 | 150 | first report take delta |
| M1 | K1=V11, K2=V11 | 10 | 210 | T1 | 10 | 210 | Start-time changed, so either ignore it or report the value as a delta  |
| M1 | K1=V11, K2=V22 | 15 | 150 | T1 | 15 | 150 | Start-time changed, so either ignore it or report the value as a delta |

SumOfSquareDev cannot be calculated without the individual updates known.

Reply:
@rghetia thank you! 

> SumOfSquareDev cannot be calculated without the individual updates known

So in this case, min/max are restarted, count/sum/avg are restarted, and SumOfSquaredDev is an unsolvable bug. This means Exporters should probably be aware of those details to export the metrics accurately. 

Would you prefer the above over `view.DisableMonotonicity(true)` configuration? 

I'm happy either way. 

Thanks!

Reply:
I prefer 
> So in this case, **min/max** are _restarted_, **count/sum/avg** are _diffed_, and SumOfSquaredDev is an unsolvable bug. This means Exporters should probably be aware of those details to export the metrics accurately.

over
> Would you prefer the above over view.DisableMonotonicity(true) configuration?

because, the later changes the meaning of aggregation as currently defined. It would be sort of a hack.

Reply:
@rghetia sounds good, then feel free to review/merge this pr :+1:

Reply:
@rghetia there's a potential new issue: how can we ensure that ExportView does _not_ get called if a `stats.Record()` was never called after the last interval? 

AFAIK, in the StatsD model, you only report to the backend when there's data. But OpenCensus will infinitely call ExportData even if the code hasn't called stats.Record since the last ReportingPeriod...right?

Reply:
In the original solution, `clearRows` makes sure that there are no rows, and therefore no new data is reported.

Reply:
> 1. Create Delta types for both Count [and Sum?] and DistributionData and implement them everywhere 

I'm not sure if that option has been discarded or not. Wouldn't this solve all issues if implemented ?